pubsub/pubsub.go

```diff
@@ -546,3 +547,135 @@ func wrapError(ec errorCoder, err error) error {
 	}
 	return gcerr.New(ec.ErrorCode(err), err, 2, "pubsub")
 }
+
+// TopicURLOpener represents types than can open Topics based on a URL.
+// The opener must not modify the URL argument. OpenTopicURL must be safe to
+// call from multiple goroutines.
+//
+// This interface is generally implemented by types in driver packages.
+type TopicURLOpener interface {
+	OpenTopicURL(ctx context.Context, u *url.URL) (*Topic, error)
+}
+
+// SubscriptionURLOpener represents types than can open Subscriptions based on a URL.
+// The opener must not modify the URL argument. OpenSubscriptionURL must be safe to
+// call from multiple goroutines.
+//
+// This interface is generally implemented by types in driver packages.
+type SubscriptionURLOpener interface {
+	OpenSubscriptionURL(ctx context.Context, u *url.URL) (*Subscription, error)
+}
+
+// URLMux is a URL opener multiplexer. It matches the scheme of the URLs
+// against a set of registered schemes and calls the opener that matches the
+// URL's scheme.
+//
+// The zero value is a multiplexer with no registered schemes.
+type URLMux struct {
+	subscriptionSchemes map[string]SubscriptionURLOpener
+	topicSchemes        map[string]TopicURLOpener
+}
+
+// RegisterTopic registers the opener with the given scheme. If an opener
+// already exists for the scheme, RegisterTopic panics.
+func (mux *URLMux) RegisterTopic(scheme string, opener TopicURLOpener) {
```

Comment:
There's so much duplicate code here and below. Any way to factor it?

I don't care if we lose some error message context (e.g. "open topic" vs. "open subscription") and I'd even be OK with some use of `interface{}` and casting.

Like, how about a general URLMux type that had methods `OpenerFor(urlstr string) (interface{}, *url.URL)` and `OpenerForURL(*url.URL) interface{}`?

Reply:
It is likely possible, but would make it harder to use (e.g., if it returned an `interface{}`, the user would have to cast it to the portable type).

If we could share implementation but leave the external interface the same, I'd be open to that, but it's hard because of all the different types.

Reply:
Yes, I meant share implementation, not change the API.

I'll approve the PR as is but can you make an issue for this and assign it to me?

Reply:
Done. https://github.com/google/go-cloud/issues/1476.